### DISCUSSION TITLE
How to access response object when there is no error

### DISCUSSION THREAD

**Discussion by dagadbm** [urql-graphql/urql#1693] [Q&A] [answered]
Hello everyone,

I currently have a header that the graphQL server I am using sends that contains a request Id.
I want to be able to log this so i can back track when I have issues.
Right now I have a generic errorExchange for this but sometimes we can also have applicational errors that are not really graphQL errors (so the  error object on [OperationResult](https://formidable.com/open-source/urql/docs/api/core/#operationresult) will be null).
How  can I access the  response of the [CombinedError](https://formidable.com/open-source/urql/docs/api/core/#combinederror) when i have no error?

**Comment by JoviDeCroock** [comment]
Generally the GraphQL spec would advise to put these into [`extensions`](https://github.com/graphql/graphql-spec/blob/main/spec/Section%207%20--%20Response.md) which would be available there.

**Comment by dagadbm** [comment]
I am still fairly new to graphQL unfortunately.

You are saying the graphQL server should put this on that extensions field then.

I will let the back-end developer know.

Stilll how would I go about getting the fetch response if its a success ? Like coding myself a custom exchange using urql?

**Reply by JoviDeCroock** [reply]
You can choose how to handle it, you can use an `exchange` and gather it from the [`operationResult`](https://github.com/FormidableLabs/urql/blob/main/packages/core/src/types.ts#L71) which to me seems like the most reliable solution but your react-hooks will also see the  extensions field. https://github.com/FormidableLabs/urql/blob/main/packages/react-urql/src/hooks/useQuery.ts#L35

**Reply by dagadbm** [reply]
i am using the core urql client btw.

I use vue2 on  my app.

So I would basically create an exchange that gets the fetch and then  pass it down so I can use it. I'll give it a try.

EDIT:

Wait Im talking about how I would go about getting the fetch object so I can access the Headers not the extensions field. That one I understand but the way the back-end server is structured the request part is all done on another layer so refactoring it is probably a non-trivial issue.  Hence my follow up question of: I understand this  is not best practice, but as a compromise, how would I go about getting  the  fetch response so I could access the headers.

**Reply by JoviDeCroock** [reply]
oh, afaik we don't really do anything with the response-headers so that might require a custom fetch-exchange 😅

**Reply by dagadbm** [reply]
can  you  point me in the direction on how i could go about doing this?

**Reply by JoviDeCroock** [reply]
You can copy from the current [fetch](https://github.com/FormidableLabs/urql/blob/main/packages/core/src/exchanges/fetch.ts) but you'll have to patch [fetchSource](https://github.com/FormidableLabs/urql/blob/f5e0b88c65c28518f0a3a8ae33ad9adff5c86ea3/packages/core/src/internal/fetchSource.ts#L28) to take response.headers in regards.

And then instead of just calling `createClient` pass in the exchanges like `exchanges: [dedupExchange, cacheExchange, yourFetchExchange]`

**Reply by dagadbm** [reply]
When you say patch you mean I will need to fork and use my own version  of urql?

Couldnt i create some other exchange that  would add the header field somewhere on the client response?

**Reply by JoviDeCroock** [reply]
I mean, that's what I'm saying, you create your own fetchExchange and use it in place of the default one. `exchanges: [dedupExchange, cacheExchange, yourFetchExchange]` is an option on the client-creation. We don't incorporate the headers by default as it's kind off not spec-compliant hence having to go out of the provided defaults. If you use `extensions` this will work just normally

**Reply by dagadbm** [reply]
ah... I see. i thought I  could use a custom exchange where I would get the fetch object and then just append it to the reply.

Something  like this:

```
// ALL exchanges should be ordered synchronous first and asynchronous last.
// reference: https://formidable.com/open-source/urql/docs/advanced/authoring-exchanges/#using-exchanges
const getHeadersFromFetch = ({ forward }) => {
  return operations$ => {
    const operationResult$ = forward(operations$);
    return operationResult$;
  };
};
const exchanges = [
  dedupExchange,
  cacheExchange,
  errorExchange({
    onError(error, operation) {
      // we are only interested in errors that came from iris
      // network errors are just noise since they can happen due to a multitude
      // of reasons such as canceled requests and timeouts
      // We only want applicational errors here to be sent into sentry.
      // Network errors datadog handles them automatically for us
      if (!error.networkError) {
        const requestId = (error.response && error.response.headers.get('X-Request-ID'))
          || 'n/a';
        Logger.error(`[IrisClient(iris-request-id ${requestId})] Error`, error.message, {
          query: print(operation.query),
          variables: JSON.stringify(operation.variables),
          errors: JSON.stringify(error.graphQLErrors),
        });
      }
    },
  }),
  getHeadersFromFetch,
  fetchExchange,
];
const client = createClient({
  exchanges,
});
``` 

But now I understand your point. 

Architectural wise do  you think  it would make sense to be able to access the fetch  response on the OperationContext type?

Because from a user perspective it is a bit weird that i get access to the fetch object on the CombinedError object but not anywhere else (at least this was surprising  for me).

I don't mind creating an issue on github for this but since you are one of the maintainers I wont do it unless you think this is something that makes sense to tackle.

Else I guess I will need to do a `pretty please with chocolate on top` to  the back-end  dev on  my team to put the request-id at the extensions key.

**Reply by JoviDeCroock** [reply]
We never expose headers as the use for responseHeaders on the client are extremely limited. Nothing is stopping you from taking the custom fetch option, it's an exchange like any other. No forks needed, I can't quite recall why we chose to expose headers on error but that might've been auth-related.

**Reply by dagadbm** [reply]
I did it  !
Omg I thought it was gonna be harder. I never did this kind of stuff before Thanks for the help!

I have two more questions:
1 - Should I just put this request-id directly in the extensions as well and at least on my side it gets spec compliant or not?

2 - There is one piece of code I dont think I can migrate and I want to know how i could do it. 
I am talking abou the [AbortController](https://github.com/FormidableLabs/urql/blob/f5e0b88c65c28518f0a3a8ae33ad9adff5c86ea3/packages/core/src/internal/fetchSource.ts#L47) .
How should i code this?

Here is my current implementation:

```javascript
import {
  make,
  filter,
  merge,
  mergeMap,
  pipe,
  share,
  takeUntil,
  onPush,
} from 'wonka';

import {
  makeFetchBody,
  makeFetchURL,
  makeFetchOptions,
} from '@urql/core/internal';

import { makeResult, makeErrorResult } from '@urql/core';

const executeFetch = (
  operation,
  url,
  fetchOptions,
) => {
  const fetcher = operation.context.fetch;

  let statusNotOk = false;
  let response;

  return (fetcher || fetch)(url, fetchOptions)
    .then(res => {
      response = res;
      statusNotOk =
        res.status < 200 ||
        res.status >= (fetchOptions.redirect === 'manual' ? 400 : 300);
      return res.json();
    })
    .then(result => {
      if (!('data' in result) && !('errors' in result)) {
        throw new Error('No Content');
      }

      return Object.assign(makeResult(operation, result, response), {
        irisRequestId: response.headers.get('X-Request-ID') || 'n/a',
      });
    })
    .catch(error => {
      if (error.name !== 'AbortError') {
        return Object.assign(makeErrorResult(operation,
          statusNotOk ? new Error(response.statusText) : error,
          response
        ), {
          irisRequestId: response.headers.get('X-Request-ID') || 'n/a',
        });
      }
    });
};

const makeFetchSource = (
  operation,
  url,
  fetchOptions,
) => {
  return make(({ next, complete }) => {
    let ended = false;

    Promise.resolve()
      .then(() => {
        if (ended) {
          return;
        }

        return executeFetch(operation, url, fetchOptions);
      })
      .then(result => {
        if (!ended) {
          ended = true;
          if (result) next(result);
          complete();
        }
      });

    return () => {
      ended = true;
    };
  });
};

export default function fetchExchange({ forward, dispatchDebug }) {
  return ops$ => {
    const sharedOps$ = share(ops$);
    const fetchResults$ = pipe(
      sharedOps$,
      filter(operation => {
        return operation.kind === 'query' || operation.kind === 'mutation';
      }),
      mergeMap(operation => {
        const { key } = operation;
        const teardown$ = pipe(
          sharedOps$,
          filter(op => op.kind === 'teardown' && op.key === key)
        );

        const body = makeFetchBody(operation);
        const url = makeFetchURL(operation, body);
        const fetchOptions = makeFetchOptions(operation, body);

        dispatchDebug({
          type: 'fetchRequest',
          message: 'A fetch request is being executed.',
          operation,
          data: {
            url,
            fetchOptions,
          },
        });

        return pipe(
          makeFetchSource(operation, url, fetchOptions),
          takeUntil(teardown$),
          onPush(result => {
            const error = !result.data ? result.error : undefined;

            dispatchDebug({
              type: error ? 'fetchError' : 'fetchSuccess',
              message: `A ${
                error ? 'failed' : 'successful'
              } fetch response has been returned.`,
              operation,
              data: {
                url,
                fetchOptions,
                value: error || result,
              },
            });
          })
        );
      })
    );

    const forward$ = pipe(
      sharedOps$,
      filter(operation => {
        return operation.kind !== 'query' && operation.kind !== 'mutation';
      }),
      forward
    );

    return merge([fetchResults$, forward$]);
  };
};
```

**Reply by JoviDeCroock** [reply]
Nice one!

Point 1 is up to you, you can choose to do with it what you like 😅 using extensions is spec-compliant yes in the future if you start using it you'd have to keep in mind to merge the back-end and manipulated extensions though.

Why wouldn't you be able to use point 2? It's basically when you have an in-flight request and you unmount that component that the request will be destroyed. The implementation can be done identical to https://github.com/FormidableLabs/urql/blob/f5e0b88c65c28518f0a3a8ae33ad9adff5c86ea3/packages/core/src/internal/fetchSource.ts#L41

**Reply by dagadbm** [reply]
Because this AbortController is a global entity right? I would I even use it outside of your codebase?
wont `typeof AbortController` return always undefined?
I searched for AbortController on your repo but I didn't really see any reference to it anywhere except [here](https://github.com/FormidableLabs/urql/blob/ef4b5b002d31e17d84fb41e5feb24d2200c9656f/packages/core/src/internal/fetchSource.test.ts#L13)

**Reply by JoviDeCroock** [reply]
AbortController is part of the browser https://caniuse.com/abortcontroller it just checks whether the current browser support it.

**Reply by dagadbm** [reply]
damn... well i had no idea about this!  TIL. im gonna mark this as resolved with my code if someone needs to do this eventually.
Thank you very much  for your support I rarely ask for it and when I do I am amazed at the kindness of everyone. Thank you :)

**Reply by kitten** [reply]
I think I'll flag this as a potential page for the docs as people will almost certainly diverge from the defacto GraphQL over HTTP spec and constraints in rare cases and we do have the internals structured for reuse

**Reply by dagadbm** [reply]
it was actually the first time I had to use internal stuff from libraries. I never stoped to think in the end  its  just a folder so there is nothing stoping me from importing wathever I want.

Its one of  the things on the [Utilities](https://formidable.com/open-source/urql/docs/api/core/#makeoperation) page for example that  i was like: wtf do I want this for? It is still not so obvious to me and I dont want to derail this any  longer  but  Im basically using graphQL and I have 0 benefits from it, no typescript, no nothing,  I  just have  to do everything by hand user voyager which the back-end dev provided. 

The ecosystem is so huge Im at a loss on what would best fit my use case.

Either way, thanks again!

**Comment by dagadbm** [comment, marked as answer]
pasting here my final version of the code (js only)

```javascript
import {
  make,
  filter,
  merge,
  mergeMap,
  pipe,
  share,
  takeUntil,
  onPush,
} from 'wonka';

import {
  makeFetchBody,
  makeFetchURL,
  makeFetchOptions,
} from '@urql/core/internal';

import { makeResult, makeErrorResult } from '@urql/core';

const executeFetch = (
  operation,
  url,
  fetchOptions,
) => {
  const fetcher = operation.context.fetch;

  let statusNotOk = false;
  let response;
  let requestId;

  return (fetcher || fetch)(url, fetchOptions)
    .then((res) => {
      statusNotOk = res.status < 200
        || res.status >= (fetchOptions.redirect === 'manual' ? 400 : 300);
      response = res;
      requestId = res.headers.get('X-Request-ID') || 'n/a';
      return res.json();
    })
    .then((res) => {
      if (!('data' in res) && !('errors' in res)) {
        throw new Error('No Content');
      }

      return Object.assign(makeResult(operation, res, response), {
        extensions: {
          ...result.extensions,
          requestId,
        },
      });
    })
    .catch((error) => {
      if (error.name !== 'AbortError') {
        return Object.assign(makeErrorResult(
          operation,
          statusNotOk ? new Error(response.statusText) : error,
          response), {
          extensions: {
            ...result.extensions,
            requestId,
          },
        });
      }
    });
};

const makeFetchSource = (
  operation,
  url,
  fetchOptions,
) => make(({ next, complete }) => {
  let ended = false;

  Promise.resolve()
    .then(() => {
      if (!ended) {
        return executeFetch(operation, url, fetchOptions);
      }
    })
    .then((result) => {
      if (!ended) {
        ended = true;
        if (result) {
          next(result);
        }
        complete();
      }
    });

  return () => {
    ended = true;
  };
});

export default function fetchExchange({ forward, dispatchDebug }) {
  return (ops$) => {
    const sharedOps$ = share(ops$);
    const fetchResults$ = pipe(
      sharedOps$,
      filter(operation => operation.kind === 'query' || operation.kind === 'mutation'),
      mergeMap((operation) => {
        const { key } = operation;
        const teardown$ = pipe(
          sharedOps$,
          filter(op => op.kind === 'teardown' && op.key === key),
        );

        const body = makeFetchBody(operation);
        const url = makeFetchURL(operation, body);
        const fetchOptions = makeFetchOptions(operation, body);

        dispatchDebug({
          type: 'fetchRequest',
          message: 'A fetch request is being executed.',
          operation,
          data: {
            url,
            fetchOptions,
          },
        });

        return pipe(
          makeFetchSource(operation, url, fetchOptions),
          takeUntil(teardown$),
          onPush((result) => {
            const error = !result.data ? result.error : undefined;

            dispatchDebug({
              type: error ? 'fetchError' : 'fetchSuccess',
              message: `A ${
                error ? 'failed' : 'successful'
              } fetch response has been returned.`,
              operation,
              data: {
                url,
                fetchOptions,
                value: error || result,
              },
            });
          }),
        );
      }),
    );

    const forward$ = pipe(
      sharedOps$,
      filter(operation => operation.kind !== 'query' && operation.kind !== 'mutation'),
      forward,
    );

    return merge([fetchResults$, forward$]);
  };
}
```
